x/mongo/driver/topology/topology.go

```diff
@@ -69,6 +69,63 @@ const (
 	SingleMode
 )
 
+type hostEnv int
+
+const (
+	genuine hostEnv = iota
+	cosmosDB
+	doumentDB
+)
+
+var (
+	cosmosDBLog   = `You appear to be connected to a CosmosDB cluster. For more information regarding feature compatibility and support please visit https://www.mongodb.com/supportability/cosmosdb`
+	documentDBLog = `You appear to be connected to a DocumentDB cluster. For more information regarding feature compatibility and support please visit https://www.mongodb.com/supportability/documentdb`
+)
+
+type envMap struct {
+	host string
+	env  hostEnv
+}
+
+type hostLogger struct {
+	logger *logger.Logger
+	envs   []envMap
+	logs   map[hostEnv]*string
+}
+
+func newHostLogger(l *logger.Logger) *hostLogger {
+	return &hostLogger{
+		logger: l,
+		envs: []envMap{
+			{".cosmos.azure.com", cosmosDB},
+			{".docdb.amazonaws.com", doumentDB},
+			{".docdb-elastic.amazonaws.com", doumentDB},
+		},
+		logs: map[hostEnv]*string{
+			cosmosDB:  &cosmosDBLog,
+			doumentDB: &documentDBLog,
+		},
+	}
+}
+
+func (l *hostLogger) log(host string) {
```

Comment:
I don't think we should have a struct for logging. IIUC the only reason we do this is to maintain state. However, it might be best to maintain this state in the code that is calling the logger, to minimize any leaking:

```go
			hostSet := map[string]bool{}
			for _, host := range parsedHosts {
				if h, _, err := net.SplitHostPort(host); err == nil {
					host = h
				}

				if !hostSet[host] {
					logTopologyEnv(t, host)
				}

				hostSet[host] = true
			}

```